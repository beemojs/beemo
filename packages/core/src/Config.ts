import { Blueprint, Predicates } from '@boost/common';
import { Configuration, createPluginsPredicate } from '@boost/config';
import { ConfigFile } from './types';

export default class Config extends Configuration<ConfigFile> {
  blueprint(predicates: Predicates): Blueprint<ConfigFile> {
    const { bool, number, object, shape, string } = predicates;

    return {
      configure: shape({
        cleanup: bool(false),
        parallel: bool(true),
      }),
      debug: bool(),
      drivers: createPluginsPredicate(predicates),
      execute: shape({
        concurrency: number(3).gt(0),
        graph: bool(true),
      }),
      module: process.env.BEEMO_CONFIG_MODULE ? string(process.env.BEEMO_CONFIG_MODULE) : string(), // .required(),
      scripts: createPluginsPredicate(predicates),
      settings: object(),
    };
  }
}
