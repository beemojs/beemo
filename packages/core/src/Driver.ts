import execa from 'execa';
import mergeWith from 'lodash/mergeWith';
import { Blueprint, isObject, optimal, Path, Predicates, predicates } from '@boost/common';
import { ConcurrentEvent, Event } from '@boost/event';
import { Plugin } from '@boost/plugin';
import {
  STRATEGY_COPY,
  STRATEGY_CREATE,
  STRATEGY_NATIVE,
  STRATEGY_NONE,
  STRATEGY_REFERENCE,
} from './constants';
import ConfigContext from './contexts/ConfigContext';
import DriverContext from './contexts/DriverContext';
import isClassInstance from './helpers/isClassInstance';
import Tool from './Tool';
import {
  Argv,
  BeemoTool,
  Driverable,
  DriverCommandOptions,
  DriverMetadata,
  DriverOptions,
  DriverOutput,
  DriverStrategy,
  Execution,
} from './types';

export default abstract class Driver<
    Config extends object = {},
    Options extends DriverOptions = DriverOptions
  >
  extends Plugin<BeemoTool, Options>
  implements Driverable {
  command: DriverCommandOptions = {};

  // Set after instantiation
  config!: Config;

  // Set after instantiation
  metadata!: DriverMetadata;

  // Set within a life-cycle
  tool!: Tool;

  output: DriverOutput = {
    stderr: '',
    stdout: '',
  };

  readonly onLoadProviderConfig = new Event<[ConfigContext, Path, Config]>('load-provider-config');

  readonly onLoadConsumerConfig = new Event<[ConfigContext, Config]>('load-consumer-config');

  readonly onMergeConfig = new Event<[ConfigContext, Config]>('merge-config');

  readonly onCreateConfigFile = new Event<[ConfigContext, Path, Config]>('create-config-file');

  readonly onCopyConfigFile = new Event<[ConfigContext, Path, Config]>('copy-config-file');

  readonly onReferenceConfigFile = new Event<[ConfigContext, Path, Config]>(
    'reference-config-file',
  );

  readonly onDeleteConfigFile = new Event<[ConfigContext, Path]>('delete-config-file');

  readonly onBeforeExecute = new ConcurrentEvent<[DriverContext, Argv]>('before-execute');

  readonly onAfterExecute = new ConcurrentEvent<[DriverContext, unknown]>('after-execute');

  readonly onFailedExecute = new ConcurrentEvent<[DriverContext, Error]>('failed-execute');

  static validate(driver: Driver) {
    const name = (isClassInstance(driver) && driver.constructor.name) || 'Driver';

    if (!isObject(driver.options)) {
      throw new Error(`\`${name}\` requires an options object.`);
    }
  }

  blueprint({ array, object, string }: Predicates): Blueprint<DriverOptions> {
    return {
      args: array(string()),
      dependencies: array(string()),
      env: object(string()),
      strategy: string(STRATEGY_NATIVE).oneOf<DriverStrategy>([
        STRATEGY_NATIVE,
        STRATEGY_CREATE,
        STRATEGY_REFERENCE,
        STRATEGY_COPY,
        STRATEGY_NONE,
      ]),
    };
  }

  bootstrap() {}

  startup(tool: Tool) {
    this.tool = tool;
    this.bootstrap();
  }

  /**
   * Special case for merging arrays.
   */
  doMerge(prevValue: unknown, nextValue: unknown): unknown {
    if (Array.isArray(prevValue) && Array.isArray(nextValue)) {
      return Array.from(new Set(prevValue.concat(nextValue)));
    }

    return undefined;
  }

  /**
   * Extract the error message when the driver fails to execute.
   */
  extractErrorMessage(error: { message: string }): string {
    return error.message.split('\n', 1)[0] || '';
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: Config): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return the module name without the Beemo namespace.
   */
  getName(): string {
    return this.name.split('-').pop()!;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return Array.isArray(this.options.args) ? this.options.args : [];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    const dependencies = Array.isArray(this.options.dependencies) ? this.options.dependencies : [];

    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Extract the current version of the installed driver via its binary.
   */
  getVersion(): string {
    const { bin, versionOption } = this.metadata;
    const version = (execa.sync(bin, [versionOption])?.stdout || '').trim();
    const match = version.match(/(\d+)\.(\d+)\.(\d+)/u);

    return match ? match[0] : '0.0.0';
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Config, next: Config): Config {
    return mergeWith(prev, next, this.doMerge);
  }

  /**
   * Handle command failures according to this driver.
   */
  processFailure(error: Execution) {
    const { stderr = '', stdout = '' } = error;
    const out = (stderr || stdout).trim();

    if (out) {
      this.setOutput('stderr', out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  processSuccess(response: Execution) {
    const { stderr = '', stdout = '' } = response;

    this.setOutput('stderr', stderr.trim());
    this.setOutput('stdout', stdout.trim());
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint<DriverCommandOptions> = {};
    const { shape, string } = predicates;

    Object.keys(options).forEach((key) => {
      blueprint[key] = shape({
        description: string().notEmpty().required(),
        type: string().oneOf<'boolean' | 'number' | 'string'>(['string', 'number', 'boolean']),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
      unknown: true,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    const { array, bool, string } = predicates;

    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[a-z]{1}[a-zA-Z0-9-]+$/u)
          .required(),
        configName: string().required(),
        configOption: string('--config'),
        configStrategy: string(STRATEGY_CREATE).oneOf([
          STRATEGY_CREATE,
          STRATEGY_REFERENCE,
          STRATEGY_COPY,
        ]),
        dependencies: array(string()),
        description: string(),
        filterOptions: bool(true),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        versionOption: string('--version'),
        watchOptions: array(string()),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }

  /**
   * Store the raw output of the driver's execution.
   */
  setOutput(type: keyof DriverOutput, value: string): this {
    this.output[type] = value.trim();

    return this;
  }
}
