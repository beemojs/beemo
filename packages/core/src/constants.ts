export const STRATEGY_NONE = 'none';

export const STRATEGY_COPY = 'copy';
export const STRATEGY_CREATE = 'create';
export const STRATEGY_REFERENCE = 'reference';
export const STRATEGY_NATIVE = 'native';
export const STRATEGY_TEMPLATE = 'template';

export const STRATEGY_BUFFER = 'buffer';
export const STRATEGY_PIPE = 'pipe';
export const STRATEGY_STREAM = 'stream';

export const KEBAB_PATTERN = /^[a-z]{1}[a-z-]+[a-z]{1}$/u;

// Keep in sync with CLI options
export const EXECUTE_OPTIONS = {
	'--concurrency': true,
	'--graph': true,
	'--workspaces': true,
};
