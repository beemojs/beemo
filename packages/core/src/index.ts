/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
