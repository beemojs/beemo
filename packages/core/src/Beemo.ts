import fs from 'fs-extra';
import path from 'path';
import { Event } from '@boost/event';
import { bool, number, string, shape } from 'optimal';
import CleanupConfigsRoutine from './routines/CleanupConfigsRoutine';
import ResolveConfigsRoutine from './routines/ResolveConfigsRoutine';
import RunDriverRoutine from './routines/RunDriverRoutine';
import Driver from './Driver';
import Context from './contexts/Context';
import DriverContext from './contexts/DriverContext';
import { Argv, Execution, BeemoConfig, BeemoPluginRegistry, DriverOptions } from './types';

export function configBlueprint() {
  return {
    configure: shape({
      cleanup: bool(false),
      parallel: bool(true),
    }),
    execute: shape({
      concurrency: number(),
      graph: bool(true),
    }),
    module: process.env.BEEMO_CONFIG_MODULE
      ? string(process.env.BEEMO_CONFIG_MODULE)
      : string().required(),
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default class Beemo<T = any> extends Tool<BeemoPluginRegistry, BeemoConfig<T>> {
  onRunDriver = new Event<[DriverContext, Driver<object, DriverOptions>]>('run-driver');

  constructor(argv: Argv, binName?: string, testingOnly: boolean = false) {
    super(
      {
        appName: 'beemo',
        appPath: path.join(__dirname, '..'),
        configBlueprint: configBlueprint(),
        configName: binName || 'beemo',
        scoped: true,
      },
      argv,
    );

    // Abort early for testing purposes
    if (testingOnly) {
      return;
    }

    this.initialize();

    // Set footer after messages have been loaded
    const footer = this.msg('app:poweredBy', { version });

    this.options.footer = `\n${this.isCI() ? '' : 'ðŸ¤–  '}${footer}`;
  }

  /**
   * Execute all routines for the chosen driver.
   */
  async runDriver(
    args: DriverContext['args'],
    driverName: string,
    parallelArgv: Argv[] = [],
  ): Promise<Execution[]> {
    const driver = this.getPlugin('driver', driverName);
    const context = this.prepareContext(new DriverContext(args, driver, parallelArgv));
    const version = driver.getVersion();

    this.onRunDriver.emit([context, driver], driverName);

    this.debug('Running with %s v%s driver', driverName, version);

    const pipeline = this.startPipeline(context)
      .pipe(new ResolveConfigsRoutine('config', this.msg('app:configGenerate')))
      .pipe(
        new RunDriverRoutine(
          'driver',
          this.msg('app:driverRun', {
            name: driver.metadata.title,
            version,
          }),
        ),
      );

    // Only add cleanup routine if we need it
    if (this.config.configure.cleanup) {
      pipeline.pipe(new CleanupConfigsRoutine('cleanup', this.msg('app:cleanup')));
    }

    return pipeline.run(driverName);
  }

  /**
   * Delete config files if a process fails.
   */
  private handleCleanupOnFailure(code: number, context: Context) {
    if (code === 0) {
      return;
    }

    // Must not be async!
    if (Array.isArray(context.configPaths)) {
      context.configPaths.forEach((config) => {
        fs.removeSync(config.path.path());
      });
    }
  }
}
