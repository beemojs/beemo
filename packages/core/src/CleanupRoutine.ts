import { Routine } from '@boost/core';
import chalk from 'chalk';
import fs from 'fs-extra';
import DriverContext from './contexts/DriverContext';
import { BeemoTool } from './types';

export default class CleanupRoutine extends Routine<DriverContext, BeemoTool> {
  bootstrap() {
    this.task(this.tool.msg('app:configCleanup'), this.deleteConfigFiles);
  }

  /**
   * Delete all temporary config files.
   */
  async deleteConfigFiles(context: DriverContext): Promise<boolean[]> {
    return Promise.all(
      context.configPaths.map(config => {
        this.debug('Deleting config file %s', chalk.cyan(config.path));

        this.tool
          .getPlugin('driver', config.driver)
          .onDeleteConfigFile.emit([context, config.path]);

        return fs.remove(config.path).then(() => true);
      }),
    );
  }
}
