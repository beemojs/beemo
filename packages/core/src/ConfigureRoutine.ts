import { Routine } from '@boost/core';
import chalk from 'chalk';
import CreateConfigRoutine from './configure/CreateConfigRoutine';
import ConfigContext from './contexts/ConfigContext';
import { BeemoTool } from './types';

export default class ConfigureRoutine<T extends ConfigContext = ConfigContext> extends Routine<
  T,
  BeemoTool
> {
  bootstrap() {
    this.resolveDependencies();
    this.setupConfigFiles();
  }

  execute(): Promise<string | string[]> {
    return this.tool.config.configure.parallel
      ? this.parallelizeRoutines()
      : this.serializeRoutines();
  }

  /**
   * Pipe a routine for every driver we need to create a configuration for,
   * and then run in parallel.
   */
  setupConfigFiles() {
    const names = [...this.context.drivers].reverse().map(driver => {
      const routine = new CreateConfigRoutine<T>(driver.name, driver.metadata.configName, {
        driver,
      });

      this.pipe(routine);

      return driver.name;
    });

    this.debug(
      'Creating config files for the following drivers: %s',
      chalk.green(names.join(', ')),
    );
  }

  /**
   * Recursively loop through an driver's dependencies, adding a dependenct driver for each,
   * starting from the primary driver (the command that initiated the process).
   */
  resolveDependencies() {
    const queue = [...this.context.drivers];

    this.debug('Resolving dependencies');

    while (queue.length > 0) {
      const driver = queue.shift()!;
      const deps = new Set(driver.getDependencies());

      this.debug('Resolving %s', driver.name);

      deps.forEach(name => {
        this.debug('  Including dependency %s', chalk.green(name));

        queue.push(this.tool.getPlugin('driver', name));
      });

      this.context.addDriverDependency(driver);
    }

    this.tool.onResolveDependencies.emit([this.context, Array.from(this.context.drivers)]);
  }
}
