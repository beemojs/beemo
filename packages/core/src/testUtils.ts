/* eslint-disable max-classes-per-file */

import path from 'path';
import execa from 'execa';
import parseArgs from 'yargs-parser';
import { mockDebugger, stubArgs, stubToolConfig } from '@boost/core/test-utils';
import Beemo from './Beemo';
import Driver from './Driver';
import Script from './Script';
import Context from './contexts/Context';
import ConfigContext, { ConfigArgs } from './contexts/ConfigContext';
import DriverContext, { DriverArgs } from './contexts/DriverContext';
import ScaffoldContext, { ScaffoldArgs } from './contexts/ScaffoldContext';
import ScriptContext, { ScriptArgs } from './contexts/ScriptContext';
import { DriverMetadata, Argv, BeemoConfig } from './types';

export { mockDebugger, stubArgs };

export class TestDriver<T extends object = {}> extends Driver<T> {
  name = 'test-driver';
}

export class TestScript<A extends object = {}, T extends object = {}> extends Script<A, T> {
  name = 'test-script';

  blueprint() {
    return {} as any;
  }
}

// Use core package since resources are located here
export const BEEMO_APP_PATH = path.join(__dirname, '..');

// Use a folder that should not cause issues / contain much code
export const BEEMO_TEST_ROOT = path.join(__dirname, '../../../tests');

export function mockTool(argv: Argv = []): Beemo {
  const tool = new Beemo(argv, '', true);

  Object.assign(tool.options, {
    appName: 'beemo',
    appPath: BEEMO_APP_PATH,
    root: BEEMO_TEST_ROOT,
    workspaceRoot: BEEMO_TEST_ROOT,
  });

  tool.config = stubToolConfig<BeemoConfig>({
    configure: {
      cleanup: false,
      parallel: true,
    },
    drivers: [],
    execute: {
      concurrency: 0,
      priority: true,
    },
    scripts: [],
  });

  return tool;
}

export function mockDriver<C extends object = {}>(
  name: string,
  tool: Beemo | null = null,
  metadata: Partial<DriverMetadata> = {},
): Driver<C> {
  const driver = new TestDriver<C>();

  driver.name = name;
  driver.tool = tool || mockTool();

  driver.setMetadata({
    bin: name,
    configName: `${name}.json`,
    title: name,
    ...metadata,
  });

  driver.bootstrap();

  return driver;
}

export function mockScript<C extends object = {}>(
  name: string,
  tool: Beemo | null = null,
): Script<{}, C> {
  const script = new TestScript<{}, C>();

  script.name = name;
  script.tool = tool || mockTool();

  return script;
}

export function applyContext<T extends Context>(context: T): T {
  context.args = parseArgs(['-a', '--foo', 'bar', 'baz']);
  context.argv = ['-a', '--foo', 'bar', 'baz'];
  context.cwd = BEEMO_TEST_ROOT;
  context.moduleRoot = BEEMO_TEST_ROOT;
  context.workspaceRoot = BEEMO_TEST_ROOT;
  context.workspaces = [];

  return context;
}

export function stubContext(): Context {
  return applyContext(new Context(stubArgs()));
}

export function stubConfigArgs(fields?: Partial<ConfigArgs>) {
  return stubArgs<ConfigArgs>({
    names: [],
    ...fields,
  });
}

export function stubConfigContext(): ConfigContext {
  return applyContext(new ConfigContext(stubConfigArgs()));
}

export function stubDriverArgs(fields?: Partial<DriverArgs>) {
  return stubArgs<DriverArgs>({
    concurrency: 1,
    live: false,
    priority: false,
    workspaces: '',
    ...fields,
  });
}

export function stubDriverContext(driver?: Driver): DriverContext {
  return applyContext(new DriverContext(stubDriverArgs(), driver || new TestDriver()));
}

export function stubScaffoldArgs(fields?: Partial<ScaffoldArgs>) {
  return stubArgs<ScaffoldArgs>({
    action: '',
    dry: false,
    generator: '',
    name: '',
    ...fields,
  });
}

export function stubScaffoldContext(
  generator: string = 'generator',
  action: string = 'action',
  name: string = '',
): ScaffoldContext {
  return applyContext(new ScaffoldContext(stubScaffoldArgs(), generator, action, name));
}

export function stubScriptArgs(fields?: Partial<ScriptArgs>) {
  return stubArgs<ScriptArgs>({
    concurrency: 1,
    name: 'foo',
    priority: false,
    workspaces: '',
    ...fields,
  });
}

export function stubScriptContext(script?: Script): ScriptContext {
  const context = applyContext(new ScriptContext(stubScriptArgs(), 'script'));

  if (script) {
    context.setScript(script, './script.js');
  }

  return context;
}

export function stubExecResult(fields?: Partial<execa.ExecaReturns>): execa.ExecaReturns {
  return {
    cmd: '',
    code: 0,
    failed: false,
    killed: false,
    signal: null,
    stderr: '',
    stdout: '',
    timedOut: false,
    ...fields,
  };
}

export function prependRoot(part: string): string {
  return path.join(BEEMO_TEST_ROOT, part);
}

export function getRoot(): string {
  return BEEMO_TEST_ROOT;
}
