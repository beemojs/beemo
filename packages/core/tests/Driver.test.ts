import Driver from '../src/Driver';
import { mockDriver, stubExecResult } from '../src/testUtils';

describe('Driver', () => {
  let driver: Driver;

  beforeEach(() => {
    driver = mockDriver('foo');
  });

  it('validates fields', () => {
    expect(() => {
      // @ts-ignore Test invalid type
      driver = new Driver({
        args: true,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  describe('formatConfig()', () => {
    it('formats to JSON if config ends in .json', () => {
      driver.metadata.configName = 'cfg.json';

      expect(
        driver.formatConfig({
          foo: 123,
          bar: 'abc',
          baz: true,
        }),
      ).toMatchSnapshot();
    });

    it('formats to JS if config ends in .js', () => {
      driver.metadata.configName = 'cfg.js';

      expect(
        driver.formatConfig({
          foo: 123,
          bar: 'abc',
          baz: true,
        }),
      ).toMatchSnapshot();
    });
  });

  describe('getArgs()', () => {
    it('returns passed fields', () => {
      driver.configure({ args: ['--foo'] });

      expect(driver.getArgs()).toEqual(['--foo']);
    });
  });

  describe('getDependencies()', () => {
    it('returns both field and metadata dependencies', () => {
      driver.metadata.dependencies = ['foo'];
      driver.configure({ dependencies: ['bar'] });

      expect(driver.getDependencies()).toEqual(['foo', 'bar']);
    });
  });

  describe('getSupportedOptions()', () => {
    it('returns an array of fields', () => {
      expect(driver.getSupportedOptions()).toEqual([]);
    });
  });

  describe('processFailure()', () => {
    let spy: jest.SpyInstance;

    beforeEach(() => {
      spy = jest.spyOn(driver.tool.console, 'logError');
    });

    afterEach(() => {
      spy.mockRestore();
    });

    it('logs stdout', () => {
      driver.processFailure(
        stubExecResult({
          stdout: 'out',
        }),
      );

      expect(spy).toHaveBeenCalledWith('out');
    });

    it('logs stderr', () => {
      driver.processFailure(
        stubExecResult({
          stderr: 'error',
        }),
      );

      expect(spy).toHaveBeenCalledWith('error');
    });

    it('logs stderr over stdout', () => {
      driver.processFailure(
        stubExecResult({
          stderr: 'error',
          stdout: 'out',
        }),
      );

      expect(spy).toHaveBeenCalledWith('error');
    });

    it('doesnt log if empty', () => {
      driver.processFailure(
        stubExecResult({
          stderr: '',
          stdout: '',
        }),
      );

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('processSuccess()', () => {
    let spy: jest.SpyInstance;

    beforeEach(() => {
      spy = jest.spyOn(driver.tool.console, 'log');
    });

    afterEach(() => {
      spy.mockRestore();
    });

    it('logs stdout', () => {
      driver.processSuccess(
        stubExecResult({
          stdout: 'out',
        }),
      );

      expect(spy).toHaveBeenCalledWith('out');
    });

    it('doesnt log stdout if empty', () => {
      driver.processSuccess(
        stubExecResult({
          stdout: '',
        }),
      );

      expect(spy).not.toHaveBeenCalledWith();
    });

    it('doesnt log stderr', () => {
      driver.processFailure(
        stubExecResult({
          stderr: 'error',
        }),
      );

      expect(spy).not.toHaveBeenCalled();
    });

    it('doesnt log if empty', () => {
      driver.processFailure(
        stubExecResult({
          stderr: '',
          stdout: '',
        }),
      );

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('mergeConfig()', () => {
    it('deep merges objects', () => {
      expect(
        driver.mergeConfig(
          {
            foo: 123,
            bar: {
              baz: 'abc',
            },
          },
          {
            bar: {
              baz: 'xyz',
              qux: true,
            },
          },
        ),
      ).toEqual({
        foo: 123,
        bar: {
          baz: 'xyz',
          qux: true,
        },
      });
    });

    it('concats arrays', () => {
      expect(driver.mergeConfig({ foo: [1, 2] }, { foo: [3, 4] })).toEqual({ foo: [1, 2, 3, 4] });
    });

    it('replaces array if new value is not an array', () => {
      expect(driver.mergeConfig({ foo: [1, 2] }, { foo: 'bar' })).toEqual({ foo: 'bar' });
    });
  });

  describe('setCommandOptions()', () => {
    it('supports description as a string', () => {
      driver.setCommandOptions({
        foo: {
          description: 'Foo',
        },
      });

      expect(driver.command).toEqual({
        foo: {
          description: 'Foo',
        },
      });
    });

    it('doesnt support description as other types', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {
            // @ts-ignore
            description: 123,
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt support empty descriptions', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {
            description: '',
          },
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('requires a description', () => {
      expect(() => {
        driver.setCommandOptions({
          foo: {},
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('setMetadata()', () => {
    const options = {
      bin: 'beemo',
      configName: 'beemo',
      title: 'Beemo',
    };

    describe('bin', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({});
        }).toThrowErrorMatchingSnapshot();
      });

      it('supports alpha, numeric, and dashes', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'foo-123',
          });
        }).not.toThrow();
      });

      it('doesnt support underscores', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'foo_123',
          });
        }).toThrow(
          'Invalid TestDriver field "bin". String does not match pattern "^[a-z]{1}[a-zA-Z0-9-]+$".',
        );
        // Broken in Node 8
        // }).toThrowErrorMatchingSnapshot();
      });

      it('doesnt support uppercase', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            bin: 'Foo_123',
          });
        }).toThrow(
          'Invalid TestDriver field "bin". String does not match pattern "^[a-z]{1}[a-zA-Z0-9-]+$".',
        );
        // Broken in Node 8
        // }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('configName', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({
            bin: 'beemo',
          });
        }).toThrowErrorMatchingSnapshot();
      });

      it('doesnt support non-strings', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            // @ts-ignore
            configName: 123,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('title', () => {
      it('is required', () => {
        expect(() => {
          driver.setMetadata({
            bin: 'beemo',
            configName: 'beemo',
          });
        }).toThrowErrorMatchingSnapshot();
      });

      it('doesnt support non-strings', () => {
        expect(() => {
          driver.setMetadata({
            ...options,
            // @ts-ignore
            title: 123,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
