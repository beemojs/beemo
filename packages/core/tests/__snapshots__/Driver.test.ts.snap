// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Driver formatConfig() formats to JS if config ends in .js 1`] = `
"module.exports = {
  \\"foo\\": 123,
  \\"bar\\": \\"abc\\",
  \\"baz\\": true
};"
`;

exports[`Driver formatConfig() formats to JSON if config ends in .json 1`] = `
"{
  \\"foo\\": 123,
  \\"bar\\": \\"abc\\",
  \\"baz\\": true
}"
`;

exports[`Driver setMetadata() bin is required 1`] = `"Invalid TestDriver field \\"bin\\". Field is required and must be defined."`;

exports[`Driver setMetadata() command options doesnt support description as other types 1`] = `"Invalid TestDriver field \\"commandOptions.foo.description\\". Must be a string, received 123."`;

exports[`Driver setMetadata() command options doesnt support empty descriptions 1`] = `"Invalid TestDriver field \\"commandOptions.foo.description\\". String cannot be empty."`;

exports[`Driver setMetadata() command options requires a description 1`] = `"Invalid TestDriver field \\"commandOptions.foo.description\\". Field is required and must be defined."`;

exports[`Driver setMetadata() configName doesnt support non-strings 1`] = `"Invalid TestDriver field \\"configName\\". Must be a string, received 123."`;

exports[`Driver setMetadata() configName is required 1`] = `"Invalid TestDriver field \\"configName\\". Field is required and must be defined."`;

exports[`Driver setMetadata() title doesnt support non-strings 1`] = `"Invalid TestDriver field \\"title\\". Must be a string, received 123."`;

exports[`Driver setMetadata() title is required 1`] = `"Invalid TestDriver field \\"title\\". Field is required and must be defined."`;

exports[`Driver validates fields 1`] = `"Invalid Driver field \\"args\\". Must be an array, received true."`;
